File: Algoritmos De Ordenamiento/Algoritmos de ordenamiento Python/quick sort.py
import time
import os
import csv

# Algoritmo de Quick Sort
def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    else:
        pivot = arr[len(arr) // 2]
        left = [x for x in arr if x < pivot]
        middle = [x for x in arr if x == pivot]
        right = [x for x in arr if x > pivot]
        return quick_sort(left) + middle + quick_sort(right)

def read_numbers_from_file(file_path):
    with open(file_path, 'r') as f:
        content = f.read()
        cleaned_content = ''.join(c for c in content if c.isdigit() or c.isspace() or c == '-')
        numbers = list(map(int, cleaned_content.split()))
    return numbers

def measure_sort_time(sort_function, arr):
    start_time = time.time()
    sort_function(arr.copy()) 
    end_time = time.time()
    return end_time - start_time

def save_results_to_csv(results, output_file):
    with open(output_file, 'w', newline='') as csvfile:
        fieldnames = ['Archivo', 'Quick Sort']
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)

        writer.writeheader()
        for file, timings in results.items():
            row = {'Archivo': file, 'Quick Sort': timings['Quick Sort']}
            writer.writerow(row)

def main():
    folder_path = r'C:\Users\saulj\Downloads\ADA_DATOS-main'  
    results = {}

    # Iterar sobre los archivos
    for filename in os.listdir(folder_path):
        if filename.endswith('.txt'):
            file_path = os.path.join(folder_path, filename)
            numbers = read_numbers_from_file(file_path)
            duration = measure_sort_time(quick_sort, numbers)
            results[filename] = {'Quick Sort': duration}

    save_results_to_csv(results, 'resultados_quick_sort.csv')

    for file, timings in results.items():
        print(f"Resultados para {file}:")
        print(f"Quick Sort: {timings['Quick Sort']:.6f} segundos")
        print()

if __name__ == "__main__":
    main()

File: Algoritmos De Ordenamiento/Algoritmos de Ordenamiento in Java/QuickSort/src/main/java/com/mycompany/quicksort/QuickSort.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 */

package com.mycompany.quicksort;
import java.io.*;
import java.util.*;
import java.util.function.Consumer;
public class QuickSort {

   // Quick Sort
    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    private static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = (low - 1);

        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, i + 1, high);
        return (i + 1);
    }

    // Swap helper function
    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    // Leer números de un archivo
    public static int[] readNumbersFromFile(String filePath) {
        List<Integer> numbers = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String line;
            while ((line = br.readLine()) != null) {
                line = line.replaceAll("[\\[\\] ]", ""); // Eliminar corchetes y espacios
                String[] nums = line.split(",");
                for (String num : nums) {
                    numbers.add(Integer.parseInt(num));
                }
            }
        } catch (IOException e) {
            System.err.println("Error al abrir el archivo: " + filePath);
            e.printStackTrace();
        }
        return numbers.stream().mapToInt(i -> i).toArray();
    }

    // Medir el tiempo de ejecución
    public static double measureSortTime(Consumer<int[]> sortFunction, int[] arr) {
        long startTime = System.nanoTime();
        sortFunction.accept(arr);
        long endTime = System.nanoTime();
        return (endTime - startTime) / 1e9; // Convertir a segundos
    }

    public static void main(String[] args) {
        String folderPath = "C:\\Users\\saulj\\Downloads\\ADA_DATOS-main"; // Ruta de la carpeta

        Map<String, Consumer<int[]>> algorithms = new HashMap<>();
        algorithms.put("Quick Sort", arr -> quickSort(arr, 0, arr.length - 1));

        Map<String, Map<String, Double>> results = new HashMap<>();

        // Obtener archivos .txt en el directorio
        File folder = new File(folderPath);
        File[] txtFiles = folder.listFiles((dir, name) -> name.toLowerCase().endsWith(".txt"));

        if (txtFiles != null) {
            for (File file : txtFiles) {
                int[] numbers = readNumbersFromFile(file.getAbsolutePath());

                // Si el archivo tiene demasiados números, saltar
                if (numbers.length > 100000) {
                    System.err.println("Advertencia: El archivo " + file.getName() + " tiene demasiados números.");
                    continue;
                }

                Map<String, Double> fileResults = new HashMap<>();
                double duration = measureSortTime(algorithms.get("Quick Sort"), Arrays.copyOf(numbers, numbers.length));
                fileResults.put("Quick Sort", duration);

                results.put(file.getName(), fileResults);
            }
        }

        // Imprimir los resultados en formato de arreglo
        System.out.println("{");
        for (Map.Entry<String, Map<String, Double>> entry : results.entrySet()) {
            System.out.print("  \"" + entry.getKey() + "\": {");
            for (Map.Entry<String, Double> algorithm : entry.getValue().entrySet()) {
                System.out.print("\"" + algorithm.getKey() + "\": " + algorithm.getValue() + ", ");
            }
            System.out.println("},");
        }
        System.out.println("}");
    }
}
